// @mazemasterjs/shared-library/Endpoint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Endpoint contains data and arguments that describe a simple service endpoint
 * This class is referenced by the Service class and contains data loaded from the
 * service.json file.
 */
class Endpoint {
    /**
     *
     * @param data Instance of the Endpoint class loaded from JSON Data (service.json)
     */
    constructor(data) {
        this.name = data.name;
        this.title = data.title;
        this.description = data.description;
        this.method = data.method;
        this.contentType = data.contentType;
        this.url = data.url;
        this.arguments = new Array();
    }
    get Name() {
        return this.name;
    }
    get Title() {
        return this.title;
    }
    get Description() {
        return this.description;
    }
    get Method() {
        return this.method;
    }
    get ContentType() {
        return this.contentType;
    }
    get Url() {
        return this.url;
    }
    get Arguments() {
        return this.arguments;
    }
    addArgument(arg) {
        this.arguments.push(arg);
    }
}
exports.Endpoint = Endpoint;
exports.default = Endpoint;
