// @mazemasterjs/shared-library/Engram.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ObjectBase_1 = require("./ObjectBase");
// An engram contains a snapshot of sensory data correlating to the
// player's position within the maze.  Each sense is a string.
//
// Example:
// sight = "You are standing in a room. There are exits to the North and South."
// sound = "You hear the sound of dripping water coming from the South."
class Engram extends ObjectBase_1.ObjectBase {
    constructor(data) {
        super();
        this.sight = '';
        this.sound = '';
        this.smell = '';
        this.touch = '';
        this.taste = '';
        if (data !== undefined) {
            this.sight = this.validateDataField('sight', data.sight, 'string');
            this.sound = this.validateDataField('sound', data.sound, 'string');
            this.smell = this.validateDataField('smell', data.smell, 'string');
            this.touch = this.validateDataField('touch', data.touch, 'string');
            this.taste = this.validateDataField('taste', data.taste, 'string');
        }
    }
}
exports.Engram = Engram;
