// @mazemasterjs/shared-library/Enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Enumerated Trophy IDs
var TROPHY_IDS;
(function (TROPHY_IDS) {
    TROPHY_IDS[TROPHY_IDS["CHEDDAR_DINNER"] = 0] = "CHEDDAR_DINNER";
    TROPHY_IDS[TROPHY_IDS["DAZED_AND_CONFUSED"] = 1] = "DAZED_AND_CONFUSED";
    TROPHY_IDS[TROPHY_IDS["DOUBLE_BACKER"] = 2] = "DOUBLE_BACKER";
    TROPHY_IDS[TROPHY_IDS["FLAWLESS_VICTORY"] = 3] = "FLAWLESS_VICTORY";
    TROPHY_IDS[TROPHY_IDS["JUMPING_JACK_FLASH"] = 4] = "JUMPING_JACK_FLASH";
    TROPHY_IDS[TROPHY_IDS["KICKING_UP_DUST"] = 5] = "KICKING_UP_DUST";
    TROPHY_IDS[TROPHY_IDS["LIGHT_AT_THE_END"] = 6] = "LIGHT_AT_THE_END";
    TROPHY_IDS[TROPHY_IDS["LOOPER"] = 7] = "LOOPER";
    TROPHY_IDS[TROPHY_IDS["MIGHTY_MOUSE"] = 8] = "MIGHTY_MOUSE";
    TROPHY_IDS[TROPHY_IDS["NERVOUS_WALK"] = 9] = "NERVOUS_WALK";
    TROPHY_IDS[TROPHY_IDS["OUT_OF_MOVES"] = 10] = "OUT_OF_MOVES";
    TROPHY_IDS[TROPHY_IDS["PAPERBACK_WRITER"] = 11] = "PAPERBACK_WRITER";
    TROPHY_IDS[TROPHY_IDS["SCRIBBLER"] = 12] = "SCRIBBLER";
    TROPHY_IDS[TROPHY_IDS["SHORTCUTTER"] = 13] = "SHORTCUTTER";
    TROPHY_IDS[TROPHY_IDS["SPINNING_YOUR_WHEELS"] = 14] = "SPINNING_YOUR_WHEELS";
    TROPHY_IDS[TROPHY_IDS["STAND_HARDER"] = 15] = "STAND_HARDER";
    TROPHY_IDS[TROPHY_IDS["TAKING_A_STAND"] = 16] = "TAKING_A_STAND";
    TROPHY_IDS[TROPHY_IDS["THE_LONGER_WAY_HOME"] = 17] = "THE_LONGER_WAY_HOME";
    TROPHY_IDS[TROPHY_IDS["THE_LONGEST_WAY_HOME"] = 18] = "THE_LONGEST_WAY_HOME";
    TROPHY_IDS[TROPHY_IDS["THE_LONG_WAY_HOME"] = 19] = "THE_LONG_WAY_HOME";
    TROPHY_IDS[TROPHY_IDS["TOO_HOT_TO_HANDLE"] = 20] = "TOO_HOT_TO_HANDLE";
    TROPHY_IDS[TROPHY_IDS["WASTED_TIME"] = 21] = "WASTED_TIME";
    TROPHY_IDS[TROPHY_IDS["WATCHING_PAINT_DRY"] = 22] = "WATCHING_PAINT_DRY";
    TROPHY_IDS[TROPHY_IDS["WISHFUL_DYING"] = 23] = "WISHFUL_DYING";
    TROPHY_IDS[TROPHY_IDS["WISHFUL_THINKING"] = 24] = "WISHFUL_THINKING";
    TROPHY_IDS[TROPHY_IDS["YOU_FELL_FOR_IT"] = 25] = "YOU_FELL_FOR_IT";
    TROPHY_IDS[TROPHY_IDS["YOU_FOUGHT_THE_WALL"] = 26] = "YOU_FOUGHT_THE_WALL";
})(TROPHY_IDS = exports.TROPHY_IDS || (exports.TROPHY_IDS = {}));
// Player States (bitwise)
var PLAYER_STATES;
(function (PLAYER_STATES) {
    PLAYER_STATES[PLAYER_STATES["NONE"] = 0] = "NONE";
    PLAYER_STATES[PLAYER_STATES["STANDING"] = 1] = "STANDING";
    PLAYER_STATES[PLAYER_STATES["SITTING"] = 2] = "SITTING";
    PLAYER_STATES[PLAYER_STATES["LYING"] = 4] = "LYING";
    PLAYER_STATES[PLAYER_STATES["STUNNED"] = 8] = "STUNNED";
    PLAYER_STATES[PLAYER_STATES["BLIND"] = 16] = "BLIND";
    PLAYER_STATES[PLAYER_STATES["BURNING"] = 32] = "BURNING";
    PLAYER_STATES[PLAYER_STATES["LAMED"] = 64] = "LAMED";
    PLAYER_STATES[PLAYER_STATES["BEARTRAPPED"] = 128] = "BEARTRAPPED";
    PLAYER_STATES[PLAYER_STATES["SLOWED"] = 256] = "SLOWED";
    PLAYER_STATES[PLAYER_STATES["DEAD"] = 512] = "DEAD";
})(PLAYER_STATES = exports.PLAYER_STATES || (exports.PLAYER_STATES = {}));
// Cardinal directions (bitwise)
// Used for movement, exits, and other direction-based functions
var DIRS;
(function (DIRS) {
    DIRS[DIRS["NONE"] = 0] = "NONE";
    DIRS[DIRS["NORTH"] = 1] = "NORTH";
    DIRS[DIRS["SOUTH"] = 2] = "SOUTH";
    DIRS[DIRS["EAST"] = 4] = "EAST";
    DIRS[DIRS["WEST"] = 8] = "WEST";
    DIRS[DIRS["LEFT"] = 16] = "LEFT";
    DIRS[DIRS["RIGHT"] = 32] = "RIGHT";
})(DIRS = exports.DIRS || (exports.DIRS = {}));
// CELL TAGS (bitwise)
var CELL_TAGS;
(function (CELL_TAGS) {
    CELL_TAGS[CELL_TAGS["NONE"] = 0] = "NONE";
    CELL_TAGS[CELL_TAGS["START"] = 1] = "START";
    CELL_TAGS[CELL_TAGS["FINISH"] = 2] = "FINISH";
    CELL_TAGS[CELL_TAGS["PATH"] = 4] = "PATH";
    CELL_TAGS[CELL_TAGS["CARVED"] = 8] = "CARVED";
    CELL_TAGS[CELL_TAGS["LAVA"] = 16] = "LAVA";
})(CELL_TAGS = exports.CELL_TAGS || (exports.CELL_TAGS = {}));
// Various Trap IDs
var CELL_TRAPS;
(function (CELL_TRAPS) {
    CELL_TRAPS[CELL_TRAPS["NONE"] = 0] = "NONE";
    CELL_TRAPS[CELL_TRAPS["PIT"] = 1] = "PIT";
    CELL_TRAPS[CELL_TRAPS["MOUSETRAP"] = 2] = "MOUSETRAP";
    CELL_TRAPS[CELL_TRAPS["TARPIT"] = 4] = "TARPIT";
    CELL_TRAPS[CELL_TRAPS["FLAMETHROWER"] = 8] = "FLAMETHROWER";
    CELL_TRAPS[CELL_TRAPS["POISON_DART"] = 16] = "POISON_DART";
    CELL_TRAPS[CELL_TRAPS["TELEPORTER"] = 32] = "TELEPORTER";
    CELL_TRAPS[CELL_TRAPS["DEADFALL"] = 64] = "DEADFALL";
    CELL_TRAPS[CELL_TRAPS["FRAGILE_FLOOR"] = 128] = "FRAGILE_FLOOR";
    CELL_TRAPS[CELL_TRAPS["CHEESE"] = 256] = "CHEESE";
})(CELL_TRAPS = exports.CELL_TRAPS || (exports.CELL_TRAPS = {}));
// Available Player Actions
var COMMANDS;
(function (COMMANDS) {
    COMMANDS[COMMANDS["NONE"] = 0] = "NONE";
    COMMANDS[COMMANDS["FACE"] = 1] = "FACE";
    COMMANDS[COMMANDS["LISTEN"] = 2] = "LISTEN";
    COMMANDS[COMMANDS["LOOK"] = 3] = "LOOK";
    COMMANDS[COMMANDS["SIT"] = 4] = "SIT";
    COMMANDS[COMMANDS["SNIFF"] = 5] = "SNIFF";
    COMMANDS[COMMANDS["STAND"] = 6] = "STAND";
    COMMANDS[COMMANDS["TURN"] = 7] = "TURN";
    COMMANDS[COMMANDS["MOVE"] = 8] = "MOVE";
    COMMANDS[COMMANDS["JUMP"] = 9] = "JUMP";
    COMMANDS[COMMANDS["WAIT"] = 10] = "WAIT";
    COMMANDS[COMMANDS["WRITE"] = 11] = "WRITE";
    COMMANDS[COMMANDS["QUIT"] = 12] = "QUIT";
})(COMMANDS = exports.COMMANDS || (exports.COMMANDS = {}));
// enumeration of possible game results
var GAME_RESULTS;
(function (GAME_RESULTS) {
    GAME_RESULTS[GAME_RESULTS["NONE"] = 0] = "NONE";
    GAME_RESULTS[GAME_RESULTS["IN_PROGRESS"] = 1] = "IN_PROGRESS";
    GAME_RESULTS[GAME_RESULTS["OUT_OF_MOVES"] = 2] = "OUT_OF_MOVES";
    GAME_RESULTS[GAME_RESULTS["OUT_OF_TIME"] = 3] = "OUT_OF_TIME";
    GAME_RESULTS[GAME_RESULTS["DEATH_TRAP"] = 4] = "DEATH_TRAP";
    GAME_RESULTS[GAME_RESULTS["DEATH_POISON"] = 5] = "DEATH_POISON";
    GAME_RESULTS[GAME_RESULTS["DEATH_LAVA"] = 6] = "DEATH_LAVA";
    GAME_RESULTS[GAME_RESULTS["WIN"] = 7] = "WIN";
    GAME_RESULTS[GAME_RESULTS["WIN_FLAWLESS"] = 8] = "WIN_FLAWLESS";
    GAME_RESULTS[GAME_RESULTS["ABANDONED"] = 9] = "ABANDONED";
})(GAME_RESULTS = exports.GAME_RESULTS || (exports.GAME_RESULTS = {}));
// enumeration of possible game states
var GAME_STATES;
(function (GAME_STATES) {
    GAME_STATES[GAME_STATES["NEW"] = 0] = "NEW";
    GAME_STATES[GAME_STATES["IN_PROGRESS"] = 1] = "IN_PROGRESS";
    GAME_STATES[GAME_STATES["FINISHED"] = 2] = "FINISHED";
    GAME_STATES[GAME_STATES["ABORTED"] = 3] = "ABORTED";
    GAME_STATES[GAME_STATES["ERROR"] = 4] = "ERROR";
})(GAME_STATES = exports.GAME_STATES || (exports.GAME_STATES = {}));
// enumeration of possible game modes
var GAME_MODES;
(function (GAME_MODES) {
    GAME_MODES[GAME_MODES["NONE"] = 0] = "NONE";
    GAME_MODES[GAME_MODES["SINGLE_PLAYER"] = 1] = "SINGLE_PLAYER";
    GAME_MODES[GAME_MODES["MULTI_PLAYER"] = 2] = "MULTI_PLAYER";
})(GAME_MODES = exports.GAME_MODES || (exports.GAME_MODES = {}));
/**
 * Function Modes - Used by some functions that need a
 * context in order to operate safely and correctly
 */
var FN_MODES;
(function (FN_MODES) {
    FN_MODES[FN_MODES["ADD"] = 0] = "ADD";
    FN_MODES[FN_MODES["REMOVE"] = 1] = "REMOVE";
})(FN_MODES = exports.FN_MODES || (exports.FN_MODES = {}));
