// @mazemasterjs/shared-library/Team.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Team is an individual code-camp team that includes a collection of Bots
 */
const Bot_1 = require("./Bot");
const ObjectBase_1 = require("./ObjectBase");
class Team extends ObjectBase_1.ObjectBase {
    constructor(data) {
        super();
        if (data !== undefined) {
            this.id = this.validateDataField('id', data.id, 'string');
            this.name = this.validateDataField('name', data.name, 'string');
            this.logo = this.validateDataField('logo', data.logo, 'string');
            this.bots = this.loadBotsArray(this.validateDataField('bots', data.bots, 'array'));
        }
        else {
            this.id = this.generateId();
            this.name = '';
            this.logo = '';
            this.bots = new Array();
        }
    }
    /**
     * Coerce json bot data into array of Bot objects and return.
     * This enforces data validation and returns concrete, functional
     * Bot objects.
     *
     * @param bots
     */
    loadBotsArray(bots) {
        const retBots = new Array();
        for (const bot of bots) {
            const newIBot = JSON.parse(JSON.stringify(bot));
            retBots.push(new Bot_1.Bot(newIBot));
        }
        return retBots;
    }
    get Id() {
        return this.id;
    }
    get Name() {
        return this.name;
    }
    set Name(teamName) {
        this.name = teamName;
    }
    get Bots() {
        return this.bots;
    }
    set Bots(bots) {
        this.bots = bots;
    }
    get Logo() {
        return this.logo;
    }
    set Logo(imageFileName) {
        this.logo = imageFileName;
    }
}
exports.Team = Team;
