// @mazemasterjs/shared-library/Trophy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ObjectBase_1 = require("./ObjectBase");
/**
 * Trohpies are awarded by the game server when certain
 * player actions or accomplisments are detected.
 *
 * See https://github.com/mazemasterjs/service-base/blob/development/data/default-trophy-list.json
 *
 */
class Trophy extends ObjectBase_1.ObjectBase {
    constructor(data) {
        super();
        this.id = this.validateDataField('id', data.id, 'string');
        this.name = this.validateDataField('name', data.name, 'string');
        this.description = this.validateDataField('description', data.description, 'string');
        this.bonusAward = this.validateDataField('bonusAward', data.bonusAward, 'number');
    }
    get Id() {
        return this.id;
    }
    get Name() {
        return this.name;
    }
    get Description() {
        return this.description;
    }
    get BonusAward() {
        return this.bonusAward;
    }
}
exports.Trophy = Trophy;
exports.default = Trophy;
